### ISSUE TITLE
help/docs to make DOS/MBR boot ISO, for PC, on Mac?

### ISSUE THREAD

**Issue by dlcarmean** [memtest86plus/memtest86plus#42]
I am trying to compare the results of this open-source code with the Passmark version, and so far I have been unable to make from the memtest.org binaries, using my Mac, a bootable USB drive.

For the images that do boot (the passmark version, and also testing with a ProxMox image) file(1) shows that they are DOS/MBR boot sector.

Nothing that I've been able to do with hdiutil and dd on MacOS Big Sur has resulted in a bootable USB stick.

Therefore, the question: can somebody update the README and/or create such an image and publish a release with that binary that I (and presumably other Mac users) can dd directly to the USB stick?

I don't presently have another system to try from linux or Windows.

Thanks.

**Comment by x86fr** [comment]
A Windows installer will be available soon (~10 days), along with the first 6.00 beta release.
An ISO of that early beta release (able to be dd'ed from MacOS or Linux) will also be available soon.

You can build it from source right now with the `make iso` command.

**Comment by Mr-Precise** [comment]
I can share my build (if possible...)
Compiled on ubuntu 20.04
![image](https://user-images.githubusercontent.com/3824793/163467457-2241a421-feb3-4fd0-9ec7-b26e9c8d9de6.png)
[memtest86plus-iso-build.zip](https://github.com/memtest86plus/memtest86plus/files/8491925/memtest86plus-iso-build.zip)

**Comment by dlcarmean** [comment]
x86fr:  my Mac build environment doesn't like the makefile out of the box, but I can look into patching:

> $ make
as -64 boot/startup.s -o boot/startup.o
clang: error: unknown argument: '-64'

Mr-Precise:  Thanks, I did get the grub version to boot, but it can't find my (usb) keyboard.

**Comment by martinwhitaker** [comment]
You could try `as --64`. GNU `as` accepts either `-64` or `--64`, but the latter now seems to be preferred.

**Comment by dlcarmean** [comment]
I'm not a C developer, haven't touched it in decades, so I'm not quite sure how to fix my env on MacOS Big Sur, but if I remove the flag altogether, I get 

../boot/startup64.S:492:9: error: invalid alignment value
        .align  4096
....

and eventually 

../boot/startup64.S:635:14: error: mach-o section specifier uses an unknown section type
        .section    ".stacks", "aw", @nobits

I'm not really sure how to proceed.

**Comment by martinwhitaker** [comment]
The obvious suggestion is to use GCC rather than clang masquerading as GCC. I don't own or use any Apple devices, so can't really help with that.

**Comment by debrouxl** [comment]
On Debian sid i386 & amd64, `AS = gcc -c -m32` / `AS = gcc -c -m64` instead of `AS = as -64` produces the expected result, and should produce the same result even on Linux distros two decades old or so:
```
$ make
gcc -c -v -m32 boot/startup.s -o boot/startup.o
Using built-in specs.
COLLECT_GCC=gcc
Target: i686-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 11.2.0-20' --with-bugurl=file:///usr/share/doc/gcc-11/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-11 --program-prefix=i686-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-targets=all --enable-multiarch --disable-werror --with-arch-32=i686 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=i686-linux-gnu --host=i686-linux-gnu --target=i686-linux-gnu
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Debian 11.2.0-20) 
COLLECT_GCC_OPTIONS='-c' '-v' '-m32' '-o' 'boot/startup.o' '-mtune=generic' '-march=i686' '-dumpdir' 'boot/'
 as -v --32 -o boot/startup.o boot/startup.s
```

```
$ make
gcc -v -c -m64 boot/startup.s -o boot/startup.o
Using built-in specs.
COLLECT_GCC=gcc
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 11.2.0-19' --with-bugurl=file:///usr/share/doc/gcc-11/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-11 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --enable-cet --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-11-uiwEqG/gcc-11-11.2.0/debian/tmp-nvptx/usr,amdgcn-amdhsa=/build/gcc-11-uiwEqG/gcc-11-11.2.0/debian/tmp-gcn/usr --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu --with-build-config=bootstrap-lto-lean --enable-link-serialization=2
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Debian 11.2.0-19) 
COLLECT_GCC_OPTIONS='-v' '-c' '-m64' '-o' 'boot/startup.o' '-mtune=generic' '-march=x86-64' '-dumpdir' 'boot/'
 as -v --64 -o boot/startup.o boot/startup.s
```

On the same platform, for testing purposes, replacing `gcc` with `clang` in the Makefile doesn't work out of the box, though: 
```
<instantiation>:10:9: error: macros cannot be nested more than 20 levels deep. Use -asm-macro-max-nesting-depth to increase this limit.
[...]
../boot/startup64.S:532:2: note: while in macro instantiation
        ptes64  0x00000000C0000000
        ^
```

**Comment by x86fr** [comment]
Binary released are now available on memtest.org.
The recommended method to burn the binary on Mac is now https://www.balena.io/etcher/

**Comment by smitty-codes** [comment]
@dlcarmean What ended up working for me was to:

1. In MacOS > open Disk Utility > Format your USB thumb drive using MS-DOS/FAT and Guid Partition
2. In my case I have an iMac Pro which has T2 chip. I needed to reboot the Mac, hold CMD+R on bootup and go into Startup Security Utility and set Secure Boot="no security" and Allow Boot Media="Allow booting from external..." (also see [apple doc](https://support.apple.com/en-us/HT208198)
3. Then download the "Linux ISO w/ Grub (64bits)" from the [website](https://www.memtest.org/), unzip the file
4. Download [Etcher](https://www.balena.io/etcher/)
5. Use Etcher to burn the unzipped ISO onto your thumb drive
6. Leave the thumb drive plugged in and reboot the Mac and hold OPTION (aka ALT) and then pick the "EFI Boot" option which will launch Grub and take you into the tool.

I also was successful using this process on an old Mac Pro 2013 (trashcan) although there is no T2 chip so step 2 can be skipped.

Hope this helps others as most people I found suggested to use Passmark's MemTest86 however I tried everything and could not get that to work. I think the key to making MemTest86+ to work was the Grub ISO version as the regular Linux and Windows ISO did NOT work for me.
